### PR TITLE
Updated the spack environment documentation

### PR DESCRIPTION
**Pull request by bvanessen**
Updated the spack environment documentation to focus on building all
of the dependencies with spack and then building LBANN, Aluminum, and
Hydrogen with the superbuild.  Fixed several typos in the external
packages.

### CODE REVIEW

File: docs/BuildingLBANN.md [hunk 1]
```diff
@@ -79,23 +79,24 @@ The following LLNL-maintained packages are optional.
...
+    correct libraries, but interferes with the way that CMake sets up
+    RPATHs.  To correctly establish the RPATH please unset the variable
+    as noted above, or you can explicity pass the RPATH fields to CMake
+    using a command such as:
 
          cmake -DCMAKE_INSTALL_RPATH=$(sed 's/:/;/g' <<< "${LIBRARY_PATH}") -DCMAKE_BUILD_RPATH=$(sed 's/:/;/g' <<< "${LIBRARY_PATH}") ...
 
-+ Build locally from source. See below for a list and descriptions of
++ Build LBANN locally from source and build Hydrogen and Aluminum
+  using the superbuild.  See below for a list and descriptions of
```
Comment (benson31) [inline docs/BuildingLBANN.md]: The code block below shows one example of building LBANN using LBANN's CMake and one example of building LBANN with the superbuild. Given this text, it should only show the latter.

File: docs/BuildingLBANN.md [hunk 2]
```diff
@@ -119,6 +120,32 @@ The following LLNL-maintained packages are optional.
...
+          -D LBANN_WITH_NCCL:BOOL=ON \
+          -D LBANN_WITH_SOFTMAX_CUDA:BOOL=ON \
+          -D LBANN_SEQUENTIAL_INITIALIZATION:BOOL=OFF \
+          -D LBANN_WITH_TBINF=OFF \
+          -D LBANN_WITH_VTUNE:BOOL=OFF \
+          -D LBANN_DATATYPE=float \
+          -D CMAKE_INSTALL_PREFIX:PATH=/path/to/lbann/install/prefix \
+          /path/to/lbann/superbuild
+
 ## Buidling with [CMake](https://cmake.org)
```
Comment (ndryden) [inline docs/BuildingLBANN.md]: Building

**Comment by bvanessen** [comment]
@ndryden Fixed the bugs that cause the incorrect version of CUDA to be detected.

**Comment by ndryden** [comment]
CMake'ing the Superbuild finished, then I got this error when actually building:
```
% cat /usr/workspace/wsa/dryden1/lbann-ndryden/build/test/lbann/stamp/LBANN-configure-err.log  
CMake Warning (dev) at build/test/lib/cmake/hydrogen/modules/FindAndVerifyMPI.cmake:11 (find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Environment variable MPI_ROOT is set to:

    /usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-rolling-release

  For compatibility, CMake is ignoring the variable.
Call Stack (most recent call first):
  build/test/lib/cmake/hydrogen/HydrogenConfig.cmake:28 (include)
  CMakeLists.txt:161 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at CMakeLists.txt:331 (find_package):
  Could not find a package configuration file provided by "CONDUIT" with any
  of the following names:

    CONDUITConfig.cmake
    conduit-config.cmake

  Add the installation prefix of "CONDUIT" to CMAKE_PREFIX_PATH or set
  "CONDUIT_DIR" to a directory containing one of the above files.  If
  "CONDUIT" provides a separate development package or SDK, be sure it has
  been installed.
```

**Comment by bvanessen** [comment]
@ndryden I have attempted to address many of your comments in the latest commit.  Please take a look.  Also @benson31 has pushed a fix for the conduit dir issue.

File: docs/BuildingLBANN.md [hunk 3]
```diff
@@ -60,66 +60,87 @@ The following LLNL-maintained packages are optional.
...
-        spack env create -d . /path/to/lbann/spack_environments/developer_release_<arch>_cuda_spack.yaml # where <arch> = x86_64 | ppc64le
+        ml gcc/7.3.1 cuda/9.2.148 spectrum-mpi/rolling-release  # Lassen / Sierra
+
++ Establish a spack environment and install software dependencies.
+  Note that there are four environments to pick from along two axis:
+  1) developers or users and 2) x86_64 and ppc64le.  For example if
+  you are a developer and want to build the inside of the git repo use
+  the following instructions:
+
+        export LBANN_HOME=<path to lbann git repo>
```
Comment (ndryden) [inline docs/BuildingLBANN.md]: You forgot the backticks.

**Comment by ndryden** [comment]
I tweaked the CMake to be this (note, in particular, no conduit):
```
cmake \
    -G Ninja \
    -D LBANN_SB_BUILD_ALUMINUM=ON \
    -D ALUMINUM_ENABLE_MPI_CUDA=ON \
    -D ALUMINUM_ENABLE_NCCL=ON \
    -D LBANN_SB_BUILD_HYDROGEN=ON \
    -D Hydrogen_ENABLE_CUDA=ON \
    -D LBANN_SB_BUILD_LBANN=ON \
    -D CMAKE_BUILD_TYPE:STRING=Release \
    -D LBANN_WITH_CUDA:BOOL=ON \
    -D LBANN_DATATYPE:STRING=float \
    -D LBANN_WITH_TOPO_AWARE:BOOL=ON \
    -D LBANN_WITH_ALUMINUM:BOOL=ON \
    -D LBANN_WITH_CUDA:BOOL=ON \
    -D LBANN_WITH_CUDNN:BOOL=ON \
    -D LBANN_WITH_NCCL:BOOL=ON \
    -D LBANN_WITH_SOFTMAX_CUDA:BOOL=ON \
    -D LBANN_SEQUENTIAL_INITIALIZATION:BOOL=OFF \
    -D LBANN_WITH_TBINF=OFF \
    -D LBANN_WITH_VTUNE:BOOL=OFF \
    -D LBANN_DATATYPE=float \
    -D CMAKE_INSTALL_PREFIX:PATH=${LBANN_INSTALL_DIR} \
    ${LBANN_HOME}/superbuild
```

I get the following build failure on Pascal in `ALUMINUM-build-out.log`:
```
[21/47] Linking CXX executable benchmark/benchmark_reductions.exe
FAILED: benchmark/benchmark_reductions.exe 
: && /usr/tce/packages/gcc/gcc-7.3.0/bin/c++  -Wall -Wextra -pedantic -g -faligned-new -O3 -DNDEBUG   benchmark/CMakeFiles/benchmark_reductions.exe.dir/benchmark_reductions.cpp.o  -o benchmark/benchmark_reductions.exe  -Wl,-rpath,/usr/workspace/wsa/dryden1/lbann-ndryden/build/test-pascal/aluminum/build/src:/usr/tce/packages/mvapich2/mvapich2-2.3-gcc-7.3.0/lib:/usr/workspace/wsa/dryden1/spack/opt/spack/linux-rhel7-x86_64/gcc-7.3.0/nccl-2.3.7-1-dbijkiqrqzelwmtiemn5k46qo352z62k/lib src/libAl.so -lcuda /usr/tce/packages/mvapich2/mvapich2-2.3-gcc-7.3.0/lib/libmpicxx.so /usr/tce/packages/mvapich2/mvapich2-2.3-gcc-7.3.0/lib/libmpi.so /usr/lib64/libhwloc.so /usr/tce/packages/gcc/gcc-7.3.0/lib64/libgomp.so -lpthread /usr/tce/packages/cuda/cuda-10.0.130/lib64/libcudart_static.a -lpthread -ldl /usr/lib64/librt.so /usr/workspace/wsa/dryden1/spack/opt/spack/linux-rhel7-x86_64/gcc-7.3.0/nccl-2.3.7-1-dbijkiqrqzelwmtiemn5k46qo352z62k/lib/libnccl.so && :
/usr/workspace/wsa/dryden1/spack/opt/spack/linux-rhel7-x86_64/gcc-7.3.0/nccl-2.3.7-1-dbijkiqrqzelwmtiemn5k46qo352z62k/lib/libnccl.so: undefined reference to `__cudaPopCallConfiguration'
collect2: error: ld returned 1 exit status
```
And similar errors for all the other binaries. I suspect this may be because I did not `ml cuda/10.0.130`, in which case the documentation should be updated to specify that this is required.

I get the following build failure on Lassen in `LBANN-build-out.log`:
```
[2/232] Building CUDA object CMakeFiles/lbann.dir/src/optimizers/sgd.cu.o
FAILED: CMakeFiles/lbann.dir/src/optimizers/sgd.cu.o 
/usr/tce/packages/cuda/cuda-9.2.88/bin/nvcc -ccbin=/usr/tce/packages/gcc/gcc-7.3.1/bin/c++ -Dlbann_EXPORTS -I. -I../../../../include -isystem=src/proto -isystem=/usr/workspace/wsa/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/protobuf-3.6.1-luswwngcm3hk7uyg7qk3kdwxqs2nbh4k/include -isystem=/usr/workspace/wsa/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/zlib-1.2.11-tz7eup7srlwdjquzfgfsyjrwt3k73jkv/include -isystem=/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-2018.12.14/include -isystem=../../install/include -isystem=../../install/include/El -isystem=../../install/include/aluminum -isystem=/usr/tce/packages/cuda/cuda-9.2.88/include -isystem=/usr/workspace/wsa/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/nccl-2.3.7-1-zrn35n6dahlg5duyo7ag4evsiusgi4tr/include -isystem=/usr/workspace/wsa/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/cub-1.7.1-3y5fhmywdrvvrebozetanelamsp2no2p/include -isystem=/usr/workspace/wsb/brain/cudnn/cudnn-7.4.2/cuda-9.2_ppc64le/include -isystem=/usr/WS1/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/opencv-3.4.3-mjyehirrzwu6dlcn7dtgkqciet2jioad/include -isystem=/usr/WS1/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/opencv-3.4.3-mjyehirrzwu6dlcn7dtgkqciet2jioad/include/opencv -isystem=/usr/workspace/wsa/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/cnpy-master-rqsopzio4qbhoir3epqbjljqa7gowrwt/include -O3 -DNDEBUG -Xcompiler=-fPIC   -pthread -arch=sm_30 -std=c++11 -x cu -c ../../../../src/optimizers/sgd.cu -o CMakeFiles/lbann.dir/src/optimizers/sgd.cu.o && /usr/tce/packages/cuda/cuda-9.2.88/bin/nvcc -ccbin=/usr/tce/packages/gcc/gcc-7.3.1/bin/c++ -Dlbann_EXPORTS -I. -I../../../../include -isystem=src/proto -isystem=/usr/workspace/wsa/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/protobuf-3.6.1-luswwngcm3hk7uyg7qk3kdwxqs2nbh4k/include -isystem=/usr/workspace/wsa/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/zlib-1.2.11-tz7eup7srlwdjquzfgfsyjrwt3k73jkv/include -isystem=/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-2018.12.14/include -isystem=../../install/include -isystem=../../install/include/El -isystem=../../install/include/aluminum -isystem=/usr/tce/packages/cuda/cuda-9.2.88/include -isystem=/usr/workspace/wsa/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/nccl-2.3.7-1-zrn35n6dahlg5duyo7ag4evsiusgi4tr/include -isystem=/usr/workspace/wsa/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/cub-1.7.1-3y5fhmywdrvvrebozetanelamsp2no2p/include -isystem=/usr/workspace/wsb/brain/cudnn/cudnn-7.4.2/cuda-9.2_ppc64le/include -isystem=/usr/WS1/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/opencv-3.4.3-mjyehirrzwu6dlcn7dtgkqciet2jioad/include -isystem=/usr/WS1/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/opencv-3.4.3-mjyehirrzwu6dlcn7dtgkqciet2jioad/include/opencv -isystem=/usr/workspace/wsa/dryden1/spack/opt/spack/linux-rhel7-ppc64le/gcc-7.3.1/cnpy-master-rqsopzio4qbhoir3epqbjljqa7gowrwt/include -O3 -DNDEBUG -Xcompiler=-fPIC   -pthread -arch=sm_30 -std=c++11 -x cu -M ../../../../src/optimizers/sgd.cu -MT CMakeFiles/lbann.dir/src/optimizers/sgd.cu.o -o CMakeFiles/lbann.dir/src/optimizers/sgd.cu.o.d
nvcc fatal   : Unknown option 'pthread'
```

**Comment by benson31** [comment]
Does the build on lassen have conduit?

**Comment by ndryden** [comment]
Unless it's on by default (which the documentation says is not the case), it should not. I used the cmake line above for both.

**Comment by ndryden** [comment]
The issue on Pascal goes away if I make sure I have not loaded CUDA or cmake in my environment before using Spack.

**Comment by benson31** [comment]
The CONDUIT-less build on lassen should now work. Working on fixing CONDUIT.

File: docs/BuildingLBANN.md (hunk 3 above)
Comment (bvanessen) [reply docs/BuildingLBANN.md]: Fixed.

**Pull request merged**